File: combinatorics/resource/ST/project_5/bai_toan_4/simple_graph/amap_to_al.cpp
#include <iostream>
#include <map>
#include <vector>
#include <sstream>

using namespace std;

// Ham chuyen ban do ke sang danh sach ke
vector<vector<int>> amapToAl(const map<int, map<int, int>>& amap, int n) {
    vector<vector<int>> al(n); // Khoi tao danh sach ke rong

    for (const auto& [i, neighbors] : amap) {
        for (const auto& [j, _] : neighbors) {
            al[i].push_back(j); // Bo qua trong so, chi lay dinh ke
        }
    }

    return al;
}

int main() {
    int n;
    cout << "Nhap so dinh: ";
    cin >> n;
    cin.ignore();

    map<int, map<int, int>> amap;
    cout << "Nhap ban do ke (dinh: (dinh,trongso)...):\n";

    for (int i = 0; i < n; ++i) {
        string line;
        getline(cin, line);
        if (line.find(":") == string::npos) continue;

        stringstream ss(line);
        string u_str, token;
        getline(ss, u_str, ':');
        int u = stoi(u_str);
        amap[u] = {};

        while (ss >> token) {
            int v, w;
            sscanf(token.c_str(), "(%d,%d)", &v, &w);
            amap[u][v] = w;
        }
    }

    // Chuyen doi va in danh sach ke
    auto al = amapToAl(amap, n);

    cout << "Danh sach ke (AL):\n";
    for (int i = 0; i < n; ++i) {
        cout << i << ": ";
        for (int v : al[i]) {
            cout << v << " ";
        }
        cout << endl;
    }

    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_4/simple_graph/amap_to_al.py
# Ham chuyen ban do ke sang danh sach ke
def amap_to_al(amap, n):
    al = [[] for _ in range(n)]  # Khoi tao danh sach ke

    for i in amap:
        for j in amap[i]:
            al[i].append(j)  # Bo qua trong so, chi lay dinh ke

    return al

# Nhap du lieu tu ban phim
n = int(input("Nhap so dinh: "))
amap = {}

print("Nhap ban do ke (dinh: (dinh,trongso)...):")
for _ in range(n):
    line = input()
    if ':' not in line:
        continue
    u_str, rest = line.split(':')
    u = int(u_str.strip())
    amap[u] = {}
    tokens = rest.strip().split()
    for token in tokens:
        v, w = map(int, token.strip("()").split(","))
        amap[u][v] = w

# Chuyen doi va in danh sach ke
al = amap_to_al(amap, n)

print("Danh sach ke (AL):")
for i, neighbors in enumerate(al):
    print(f"{i}: {neighbors}")